dnsflow_read.py

```diff
@@ -1,125 +1,179 @@
-                continue
-            for record in pkt['data']:
-                yield ts, record
-
-    # Iterate over dnsflow pkts.
-    def pkt_iter(self):
-        while 1:
-            rv = self._pcap.next()
-            if rv == None:
-                if self.pcap_file is not None:
-                    # eof
-                    break
-                else:
-                    # interface, hit to_ms
-                    continue
-            pktlen, buf, ts = rv
-            pkt, err = process_pkt(self._pcap.datalink(), ts, buf,
-                    stats_only=self.stats_only)
-            if err is not None:
-                print err
-                continue
-            yield pkt
-
-#
+            self.pcap = pcap.pcap(name=interface, snaplen=SNAPLEN, promisc=True, timeout_ms=TIMEOUT)
+            self.pcap.setfilter(self.pcap_filter, optimize=1)
+
+    def iter_interface(self):
+        """ Live-capture packets and process them. """
+        print('Listening on %s: %s' % (self.pcap.name, self.pcap.filter))
+        while True:
+            next_item = next(self.pcap)
+            if next_item is not None:
+                ts, frame = next_item
+                # When live-capturing, have already set the filter, so no need to do additional filtering
+                res = self.handle_frame(ts, frame)
+                if res:
+                    yield res
+
+    def iter_pcap(self):
+        """ Iterate through pcap file and process packets. """
+        for ts, frame in self.pcap:
+            res = self.handle_frame(ts, frame, filter=True)
+            if res:
+                yield res
+
+    def handle_frame(self, ts, frame, filter=False):
+        pkt, err = process_pkt(
+            self.pcap.datalink(), ts, frame, stats_only=self.stats_only, filter=filter
+        )
+        if err is not None:
+            print(err)
+            pkt = None
+
+        return pkt
+
+
+def dnsflow_lo(lo):
+    """ Check if dpkt.loopback.Loopback represents DNSFlow """
+    dnsflow = False
+    if lo.family == socket.AF_INET:
+        ip = lo.data
+        if type(ip.data) == UDP and ip.data.dport == DNSFLOW_PORT:
+            dnsflow = True
+
+    return dnsflow
+
+
+def dnsflow_eth(eth):
+    """ Check if dpkt.loopback.Loopback represents DNSFlow """
+    dnsflow = False
+    if eth.type != dpkt.ethernet.ETH_TYPE_IP:
+        return dnsflow
+
+    ip = eth.data
+    if ip.p == IP_PROTO_UDP and ip.data.dport == DNSFLOW_PORT:
+        dnsflow = True
+
+    return dnsflow
+
+
 # Returns a tuple(pkt_contents, error_string).
 # error_string is None on success; on failure it contains a message
 # describing the error.
 # pkt_contents is a dict containing the unmarshaled data from the packet. It
 # may be incomplete or empty on error.
 # stats_only - set to True to parse stats pkts and headers only of data pkts.
-def process_pkt(dl_type, ts, buf, stats_only=False):
+def process_pkt(dl_type, ts, buf, stats_only=False, filter=False):
```

Comment:
this is a really long and messy function that half of me really wanted to refactor (and that pycharm hated), but i don't think its worth taking the extra time if i'm being honest with myself. so changed below are limited to adding the default filtering for just dnsflow packets (for when in the pcap file case), and python3 changes